### ISSUE TITLE
Task Dialog: Improve hyperlink usage

### ISSUE THREAD

**Issue by kpreisser** [dotnet/winforms#1199]
:arrow_double_down::arrow_double_down: [**CLICK HERE TO SCROLL DOWN TO THE FINAL PROPOSAL**](https://github.com/dotnet/winforms/issues/1199#issuecomment-822142698) :arrow_double_down::arrow_double_down:

----

Hi,

as requested by @RussKie in https://github.com/dotnet/winforms/issues/146#issuecomment-502493259, I've opened a new issue for this topic.
**Note:** The listed options don't need to change the currently proposed API in an incompatible way, so this issue can be addressed at a later time.

The native Windows Task Dialog allows to use hyperlinks in certain text elements (text/content, footer, expander) by setting the `TDF_ENABLE_HYPERLINKS` flag. This allows you to specify a link with the form `<a href="mylink">text</a>` (similar to HTML). When the user clicks on the link, the `TDN_HYPERLINK_CLICKED` is sent to the callback with the content of the `href` attribute as parameter.

This is reflected in the currently proposed API in #146: To use links, you set the `TaskDialogPage.HyperlinksEnabled` property to `true`, and then you can set e.g. 
```c#
page.Text = "This text contains <a href=\"link1\">a link</a>.";
```
to show a link in the task dialog text:

![grafik](https://user-images.githubusercontent.com/13289184/59568278-dbd2a080-9078-11e9-9d76-032795ffbbc8.png)

Then, you have to handle the `TaskDialogPage.HyperlinkClicked` event to get notified when the user clicks a link. That event will provide you the text of the `href` attribute, in that case `link1`.

However, there are some problems with this approach (as is simply a mapping of the native API):
* Users need to know the exact syntax to specify a link (which is similar to HTML), and in source code you need to escape the double quotes.
* There is a risk that when you enable hyperlinks and then show strings from external sources, they might get misinterpreted as hyperlink even if you actually just want it to show as plain text. As noted in the documentation for [`TDF_ENABLE_HYPERLINKS`](https://docs.microsoft.com/en-us/windows/desktop/api/commctrl/ns-commctrl-_taskdialogconfig#tdf_enable_hyperlinks), this could cause security vulnerabilities, for example if you did something like `Process.Start(e.Hyperlink)` in the `HyperlinkClicked` event.
(A safer way is not to use links like `https://...` directly in the `href`, but instead use keys like `linkA`, and then in the event, compare the key to open the corresponding link.)
* In order to prevent strings from external sources to be interpreted as links (or when you actually want to display a string like `<a href="..."></a>` as part of the text directly without interpreting as hyperlink), you need to replace `"<a"` e.g. with `"<\u200Ba"` (zero width space), so that the task dialog interprets `<` as literal character.
A minor downside of this is that if you then copy the dialog's text contents with <kbd>Ctrl</kbd>+<kbd>C</kbd>, that invisible character is also copied.
(I didn't find another way for this, as the HTML/XML-like `&lt;` doesn't seem to work; and while replacing `"<a"` with `"<<a></a>a"` (which I thought previously) would fix the copy problem with the invisible space, it cannot be used as it creates empty hyperlinks that are reachable by pressing <kbd>Tab ↹</kbd>.)
* When you enable `HyperlinksEnabled` and there is at least one link (`<a ...>`) present in the text, an `&` character will be interpreted to specify a mnemonic char, instead of displaying it directly (which is the behavior when `HyperlinksEnabled` is not set), although then actually nothing happens when you press <kbd>Alt</kbd> and the character.
Therefore, you would generally need to replace `&` with `&&` in the strings when `HyperlinksEnabled` is set and you have at least one link in the text.
(But if you then copy the task dialog's text contents with <kbd>Ctrl</kbd>+<kbd>C</kbd>, it will copy the escaped form (`&&`) instead of the displayed form (`&`).)

## Option A

As suggested by @GSPP:
Don't change the existing API, but add new methods that can be used to retrieve strings with the correct syntax when `EnableHyperlinks` is set, for example::

```c#
    public static string GetHyperlinkText(string href, string text)
    {
        // Don't allow "</a>" in the link text because we cannot escape that.
        if (text.Contains("</a>", StringComparison.OrdinalIgnoreCase))
            throw new ArgumentException();
        if (href.Contains("\"", StringComparison.OrdinalIgnoreCase))
            throw new ArgumentException();

        return $"<a href=\"{href}\">{text.Replace("&", "&&")}</a>";
    }

    public static string GetNonHyperlinkText(string text)
    {
        return text.Replace("&", "&&").Replace("<a", "<\u200Ba").Replace("<A", "<\u200BA");
    }
```

Then, you could set the text like this:
```c#
    page.Text = TaskDialog.GetHyperlinkText("myLink", "Hello World & others!") +
        TaskDialog.GetNonHyperlinkText(" External text: <a href=\"test\">&</a>");
```
![grafik](https://user-images.githubusercontent.com/13289184/59962698-2f197880-94e9-11e9-8dc0-744539da8206.png)

With the helper methods, you wouldn't need to specify the exact syntax for the link and you wouldn't need to escape special characters like `<` by yourself.

## Option B

As suggested by @jnm2: Also don't change the existing API but add a new method
```c#
    public static string FormatWithHyperlinks(FormattableString str)
```
on the `TaskDialog` which you then could call with an interpolated string, that takes the `href` as argument and the link text as format string (or the other way round):
```c#
    page.Text = TaskDialog.FormatWithHyperlinks(
        $"Some text, then {"https://foo/":a link} & {"https://bar/":another link}; <a href=\"x&y\">no link</a>");

    // Or:
    page.Text = TaskDialog.FormatWithHyperlinks(
        $"Some text, then {"a link":https://foo/} & {"another link":https://bar/}; <a href=\"x&y\">no link</a>");
```

![grafik](https://user-images.githubusercontent.com/13289184/59965143-8419b680-950a-11e9-9bb6-c176d66c5cd4.png)

(However, with this option there will probably need to be another method that allows to escape strings (like `GetNonHyperlinkText()` in Option A) if you then want to embed other strings which shouldn't get interpreted as link.)

Thanks!

**Comment by jnm2** [comment]
One thing to be aware of with FormattableString. If you manually create a variable and don't explicitly type it as FormattableString, C# prefers the type `string` which means the templated string is rendered to `string` before getting passed.

```cs
var url1 = "https://foo";
var url2 = "https://bar";
var text = $"Some text, then {url1:a link}, then {url2:another link}." 
// text now holds merely the string "Some text, then t, then g."
TaskDialog.FormatWithHyperlinks(text);
```

A Roslyn analyzer could provide a warning and a fix for this scenario.

**Comment by jnm2** [comment]
One thing to be aware of with FormattableString. If you manually create a variable and don't explicitly type it as FormattableString, C# prefers the type `string` which means the templated string is rendered to `string` before getting passed.

```cs
var url1 = "https://foo";
var url2 = "https://bar";
var text = $"Some text, then {url1:a link}, then {url2:another link}." 
// text now holds merely the string "Some text, then t, then g."
```

This is okay so long as there is no overload of `FormatWithHyperlinks` that accepts a string. The compiler will draw your attention to it by erroring:

```cs
// CS1503 Argument 1: cannot convert from 'string' to 'System.FormattableString'
TaskDialog.FormatWithHyperlinks(text);
```

**Comment by jnm2** [comment]
It doesn't appear to be possible to manually create and handle link clicks. This is something we'll need before we can move away from our custom implementation. Is this issue tracking the ability to respond to link clicks, or should I open a new one?

**Comment by RussKie** [comment]
We have completely descoped hyperlink-related implementations due to an observed confusion during our user studies with the proposed implementation. Whilst the API followed the Win32 implementation, they felt somewhat foreign to Windows Forms users.

To add these API we need to design a new set of API that feel "more native" to Windows Forms.

**Comment by jnm2** [comment]
@RussKie Is there enough information on it that the community can help with that, or are we blocked waiting for an internal team?

**Comment by RussKie** [comment]
This piece of work isn't high on our priority list, so a community contribution is quite welcome. I'm happy to work with the community and help to see it through.

**Comment by jnm2** [comment]
That's great! Are you looking for API ideas, or do already have some concept of the shape it should be? If the former, do you have notes on the pitfalls that were discovered that should be avoided while looking for ideas?

**Comment by RussKie** [comment]
I've looked through the notes, and appears @kpreisser has captured them very well in the OP (all that before Option A and B).

I haven't had a chance to spend any meaningful time pondering on API (busy getting .NET 5.0 ready), but I don't think either of the proposed options will meet the bar in their current shape.
The TD API has undergone a number revisions from the original proposal, and we'll need to come up with an API that will fit in with the final design, ideally with a minimal learning curve for a general Windows Forms developer.

**Comment by jnm2** [comment]
@RussKie How about an API roughly like this?

```cs
// No need to specify href since it's basically just a note to yourself in the form of a string tag.

var firstLink = new TaskDialogHyperlink("a link");
firstLink.Click += FirstLink_Click;

var secondLink = new TaskDialogHyperlink("another link") { Tag = "some info" };
secondLink.Click += HandlerThatMakesUseOfTheTag;

page.TextContent.Add("Plain text (automatically escaped for you if necessary) followed by ");
page.TextContent.Add(firstLink );
page.TextContent.Add(", and then ");
page.TextContent.Add(secondLink);
page.TextContent.Add(".");
```

It's less winformsy, but maybe to really drive home the point that you're on the hook to handle the click (no automatic link handling), the TaskDialogHyperlink constructor requires both `string displayText` and `Action clickHandler`.

**Comment by nikeee** [comment]
I think it's less practical if one has to translate the text of the dialog into multiple languages. The order of the items might change or could even get split up.

**Comment by jnm2** [comment]
@nikeee What would work best for translations, having format strings with required tags?

`Plain text (automatically escaped for you if necessary) followed by {Tag1:a link}, and then {Tag2:another link}.`

**Comment by jnm2** [comment]
New proposal because I agree that breaking into segments makes localization painful.

```cs
var firstLink = new TaskDialogHyperlink("a link");
firstLink.Click += FirstLink_Click;

var secondLink = new TaskDialogHyperlink("another link") { Tag = "some info" };
secondLink.Click += HandlerThatMakesUseOfTheTag;

var page = new TaskDialogPage
{
    HyperlinkText = new TaskDialogHyperlinkText(
        Resources.SomeMessage, // "Plain auto-escaped text with {0:link A}, an injected value ({1}), and {2:link B}."
        firstLink,
        42,
        secondLink)
};
```

**Comment by jnm2** [comment]
However, there will probably people who want to build the text programmatically by segments (I've done this recently), and they shouldn't have to create a format string in order to do so. Perhaps it would be better to blend these two, basically taking the first one and adding a `TextContent.Add(string format, params object[] args)` overload.

**Comment by RussKie** [comment]
Thank you! Keep the ideas coming. I'll set some time aside this week to think about this too.

@kpreisser would love to hear you thoughts too, as well as feasibility from the API perspective.

**Comment by RussKie** [comment]
I'm flat out busy with prep'ing the 5.0 release, and I don't have any capacity for this until we ship. Sorry for the delay.

**Comment by RussKie** [comment]
Apologies for taking more time than I promised. Thank you to everyone who took their time to share their thoughts.
Here are my thoughts.

I don't think the **Option A** would work well because it puts a lot of responsibility on a developer, which in turn makes it rather error-prone. Nonetheless,  @GSPP thank you for your proposal.

The **Option B** has more potential, though we need to iterate on it more.

@jnm2 I like the idea of `TaskDialogHyperlink`, and agree with your assessment of discoverability and clarity of intent:

> ```cs
> var secondLink = new TaskDialogHyperlink("link text") { Tag = "some info" };
> secondLink.Click += HandlerThatMakesUseOfTheTag;
> ```
> 
> It's less winformsy, but maybe to really drive home the point that you're on the hook to handle the click (no automatic link handling), the TaskDialogHyperlink constructor requires both string displayText and Action clickHandler. On the other hand, this benefit could still be had with the Click event if there is an in-the-box analyzer that produces a warning if you create a TaskDialogHyperlink without handling the Click event.

This API should also fit in nicely along the existing TaskDialog API. 

I do agree with @nikeee that breaking up a string makes localisation-related work harder, so we may want to avoid this. At the same time this API are is not dissimilar to the `StringBuilder` API :).

I like the proposal to use text with placeholders (e.g. `Plain text (automatically escaped for you if necessary) followed by {0}, and then {1}.`).
And if `TaskDialogHyperlink` can be converted to a string, it would work similar to this:
![image](https://user-images.githubusercontent.com/4403806/101731215-6af63280-3b0f-11eb-8b13-4eb6ca2b38ce.png)

```cs

TaskDialogButton result = TaskDialog.ShowDialog(this, new TaskDialogPage()
{
    EnableHyperlinks = true,
    Text = string.Format(fmt, link1, link2),
}
```

As @kpreisser suggested earlier, handling a click event we'd pass the link ID (or whatever we end up calling it) - being "link1" and "link2" from the sample, so a user can handle it explicitly.
This way we'd address use cases where a link is a hyperlink, or a custom in-app link (e.g. click a link to open another dialog in the app).


What do you think?

**Comment by jnm2** [comment]
@RussKie If TaskDialogHyperlink has a Click event or e.g. `Action onClick` constructor parameter, then the TaskDialogHyperlink instance itself will have to be added to the task dialog page somehow and not just the text. Then `EnableHyperlinks` would not need to be a property because it could be determined by the presence of a TaskDialogHyperlink in the page.

If TaskDialogHyperlink doesn't have a Click event, then I like `LinkId` and `LinkText`, and I guess there would be a HyperlinkClick event on the page/dialog, and then there would be no way to avoid the EnableHyperlinks property since links are created by setting `Text` rather than by adding TaskDialogHyperlink instances to the page.

My preference is for the former, where `string.Format` is not used for localization and instead something like `TextContent.Add(string format, params object[] args)` is used that allows TaskDialogHyperlink instances to be added to the page and have Click events or `Action onClick`-style constructor parameters.

The former also allows code to programmatically examine the text content and reason about hyperlinks easily without having to parse `<a` syntax.

**Comment by jnm2** [comment]
@RussKie Also note that `LinkText` will almost always need to be localized when any of the text is localized.

**Comment by RussKie** [comment]
We had a brainstorm session with @KlausLoeffelmann and @dreddy-work, exploring various scenarios and developer experience.

I've looked at the original @kpreisser's proposal (through the ILSpy in the old build I happen to have lying around), and the hyperlink click event was implemented like this (a snip from a larger piece):
```cs
private int HandleTaskDialogCallback(IntPtr hWnd, Interop.TaskDialog.TASKDIALOG_NOTIFICATIONS notification, IntPtr wParam, IntPtr lParam)
{
    switch (notification)
    {
		case Interop.TaskDialog.TASKDIALOG_NOTIFICATIONS.TDN_HYPERLINK_CLICKED:
		{
			string link = Marshal.PtrToStringUni(lParam);
			TaskDialogHyperlinkClickedEventArgs eventArgs = new TaskDialogHyperlinkClickedEventArgs(link);
			_boundPage.OnHyperlinkClicked(eventArgs);
			break;
		}
    }
}
```

From Win32 we're only getting the _hyperlink_ of the link clicked (or "executablestring" as the [docs](https://docs.microsoft.com/windows/win32/api/commctrl/ns-commctrl-taskdialogconfig) call it), and whilst it feels logical to place `Click` event at a `TaskDialogHyperlink` type, in reality it is handled at the `TaskDialogDialog` level.
Of course with some plumbing we can maintain a collection of hyperlinks objects, and then raise an event at the original instance, but I fail to see a significant value that justifies all that added complexity. (We actually went down this road in our discussions, failing to check the Win32 API.)

With that the proposed `TaskDialogHyperlink` type is really a convince type, and strictly speaking it is not necessary for the hyperlinks functionality, but it makes it easier for users to render hyperlink markup:
```cs
public class TaskDialogHyperlink
{
    public TaskDialogHyperlink(string text, string hyperlink) { ... }   
    public string Text { get; }
    public string Hyperlink { get; }
    public override string ToString() => /* render <A HREF="Hyperlink">Text</A> with necessary escaping */
}
```
The rest is plumbing for the most part:
```cs
public class TaskDialogHyperlinkClickedEventArgs : EventArgs
{
	internal TaskDialogHyperlinkClickedEventArgs(string hyperlink)
	{
		Hyperlink = hyperlink;
	}

	/// <summary>
	/// The value of the <c>href</c> attribute of the hyperlink that the user clicked.
	/// </summary>
	/// <remarks>
	/// Please note: In order to avoid possible security vulnerabilities when showing content
	/// from unsafe sources in a task dialog, you should always verify the value of this
	/// property before actually opening the link.
	/// </remarks>
	public string Hyperlink { get; }
}

public class TaskDialogPage
{
    protected internal void OnHyperlinkClicked(TaskDialogHyperlinkClickedEventArgs e)
    {
        HyperlinkClicked?.Invoke(this, e);
    }

    internal void Bind(TaskDialog owner, out Interop.TaskDialog.TASKDIALOG_FLAGS flags, ...)
    {
        // ...

        // Set EnableHyperlinks based on whether there is a wired handler
        EventHandler hyperlinkClicked = HyperlinkClicked;
        SetFlag(Interop.TaskDialog.TASKDIALOG_FLAGS.TDF_ENABLE_HYPERLINKS, hyperlinkClicked  is null);

        // ...
    }
}

public class TaskDialogDialog
{
    private int HandleTaskDialogCallback(IntPtr hWnd, Interop.TaskDialog.TASKDIALOG_NOTIFICATIONS notification, IntPtr wParam, IntPtr lParam)
    {
        switch (notification)
        {
            case Interop.TaskDialog.TASKDIALOG_NOTIFICATIONS.TDN_HYPERLINK_CLICKED:
            {
                string hyperlink = Marshal.PtrToStringUni(lParam);
                TaskDialogHyperlinkClickedEventArgs eventArgs = new TaskDialogHyperlinkClickedEventArgs(hyperlink);
                _boundPage.OnHyperlinkClicked(eventArgs);
                break;
            }
        }
    }
}
```




> If TaskDialogHyperlink doesn't have a Click event, then I like `LinkId` and `LinkText`, and I guess there would be a HyperlinkClick event on the page/dialog, and then there would be no way to avoid the EnableHyperlinks property since links are created by setting `Text` rather than by adding TaskDialogHyperlink instances to the page.

I don't think this is a problem. As shown above, we can infer and set the state of `EnableHyperlinks` based on whether `TaskDialogPage.HyperlinkClicked` event is wired or not.

> My preference is for the former, where `string.Format` is not used for localization and instead something like `TextContent.Add(string format, params object[] args)` is used for localization.

I'm not sure I understand the issue here. 
I don't believe it is uncommon to have a localisable content with placeholders, e.g. "_This sting will show {0} text_". Each instance of hyperlink can have their own localisable content used for `Text` property, and then at runtime all of these localised resources come together via `string.Format` or string interpolation. 
This approached is used in the Windows Forms SDK, and pretty much the same is employed in Git Extensions project.

**Comment by GSPP** [comment]
>     public override string ToString() => /* render <A HREF="Hyperlink">Text</A> with necessary escaping */

Is it really a good idea to make the `ToString` output HTML? I'd make it a nice debug representation that is useful in the debugger. Then, there should be a separate method `GetHTML` (or no such public functionality).

**Comment by RussKie** [comment]
> Then, there should be a separate method `GetHTML` (or no such public functionality).

Without it we lose string interpolation abilities, e.g. this won't be possible `page.Text = $"Click my {link1}";`, and we'll have something like `page.Text = $"Click my {link1.AsHtml()}";`

**Comment by jnm2** [comment]
> As shown above, we can infer and set the state of `EnableHyperlinks` based on whether `TaskDialogPage.HyperlinkClicked` event is wired or not.

This worries me. I might want to have a helper method that always wires up HyperlinkClicked without knowing whether the text was intended to be interpreted as having hyperlinks. There is not a strong enough dichotomy between regular text that just happens to fall into the link pattern and an actual intended hyperlink.

**Comment by jnm2** [comment]
Can `TaskDialogHyperlink.Hyperlink` be renamed `Id` or `Tag` to show that there is no interpretation going on that requires the href ID to be a URL or link in any sense?

**Comment by KlausLoeffelmann** [comment]
> This worries me. I might want to have a helper method that always wires up HyperlinkClicked without knowing whether the text was intended to be interpreted as having hyperlinks. There is not a strong enough dichotomy between regular text that just happens to fall into the pattern of a link (maybe it's quoting user input!) and an actual intended hyperlink.

Wouldn't that be a problem in the native Win32 API already?

**Comment by jnm2** [comment]
@KlausLoeffelmann You do need to specifically enable hyperlinks in the Win32 API using `TDF_ENABLE_HYPERLINKS`. I would hope that we would not see it as a particular goal to have the object model resemble the Win32 API though. The way the Win32 API requires hyperlinks to be encoded into the text is something I was hoping the Windows Forms object model would not replicate.

**Comment by KlausLoeffelmann** [comment]
So, you mean, only if there was a definition of HyperLinks (as objects), you would implicitly have the TDF_ENABLE_HYPERLINKs flag set?

**Comment by KlausLoeffelmann** [comment]
Let me just throw out a piece of code (based on a stage we had it yesterday in the middle of a discussion). Not as a statement of intent to do it that way - just to see if the thought goes that way. At least not yet. 😄 

```
namespace TaskDialogApiDiscussion
{
    public class TaskDialogHyperlink
    {
        public TaskDialogHyperlink(string text, string identifier)
        {
            Text = text;
            Identifier = identifier;

            // Not really sure, if this is necessary.
            //HyperlinkId = Guid.NewGuid();
        }

        public string Text { get; }
        public string Identifier { get; }

        public override string ToString()
        {
            return Identifier;
        }
    }

    public static class TaskDialogExtension
    {
        public static TaskDialogHyperlink CreateHyperlink(this TaskDialogPage page, string text, string url)
        {
            var hyperlink = new TaskDialogHyperlink(text, url);
            // page.Hyperlinks.Add(hyperlink);
            return hyperlink;
        }
    }

    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private void button1_Click(object sender, EventArgs e)
        {
            var page = new TaskDialogPage();

            // Hyperlink doesn't have a public constructor, so we'd always ensure we know
            // the hyperlinks because they will be bound to the page.
            TaskDialogHyperlink hyperlink = page.CreateHyperlink(text: "Hyperlinktext", url: "URL");
            hyperlink.Click += SomeMethod;

            // {hyperlink} NOW enables TDF_ENABLE_HYPERLINKS, because we can match the hyperlink (object).
            page.Text = $"This is my {hyperlink}";
            var taskDialog = TaskDialog.ShowDialog(page);
        }
    }
}
```

**Comment by RussKie** [comment]
> > Of course with some plumbing we can maintain a collection of hyperlinks objects, and then raise an event at the original instance, but I fail to see a significant value that justifies all that added complexity.
> 
> The value I see is that there would be an object model at the right conceptual level with full safety and intentionality, rather than a dual state constructed-text-as-data-model Text property.

This would be at the expense of a) memory consumption, b) more CPU cycles, and c) code maintainability. All of which IMO are costly.

From architectural perspective, there's a single channel to receive notifications for **all** hyperlinks on the page, that get clicked. We have to maintain a collection of hyperlinks, and an associated collection of handler instances, and route a click notification to the hyperlink instance that it was wired to.
A lot of moving parts for an arguable convenience.

> Can `TaskDialogHyperlink.Hyperlink` be renamed `Id` or `Tag`

Sure, this isn't set in stone yet, and can be discussed. I've chosen "hyperlink" though "url" is probably more appropriate here. The [docs](https://docs.microsoft.com/windows/win32/api/commctrl/ns-commctrl-taskdialogconfig) refer to it as `executablestring`.

**Comment by RussKie** [comment]
> This worries me. I might want to have a helper method that always wires up HyperlinkClicked without knowing whether the text was intended to be interpreted as having hyperlinks. There is not a strong enough dichotomy between regular text that just happens to fall into the pattern of a link (maybe it's quoting user input!) and an actual intended hyperlink.

In this case: `page.EnableHyperlinks = true` :)

**Comment by KlausLoeffelmann** [comment]
> This would be at the expense of a) memory consumption, b) more CPU cycles, and c) code maintainability. All of which IMO are costly.

Is this relevant here? I mean, it's not that a TaskDialog has 10,000 pages or is called 100,000 times per second.

I simply find the solution we came up with yesterday still more elegant and intuitive.

**Comment by jnm2** [comment]
> This would be at the expense of a) memory consumption, b) more CPU cycles, and c) code maintainability. All of which IMO are costly.

I agree with Klaus that the CPU cycles and memory consumption seem unlikely to be an issue.

> From architectural perspective, there's a single channel to receive notifications for all hyperlinks on the page, that get clicked. We have to maintain a collection of hyperlinks, and an associated collection of handler instances, and route a click notification to the hyperlink instance that it was wired to.
> A lot of moving parts for an arguable convenience.

The same argument could be made about the architecture of WndProc. Windows Forms has not previously had it as a goal to expose the underlying Win32 API architecture.

**Comment by RussKie** [comment]
```cs
const string MyUrl = "URL";

TaskDialogHyperlink hyperlink1 = page.CreateHyperlink(text: "Click this link", url: MyUrl);
hyperlink1.Click += SomeMethod;

TaskDialogHyperlink hyperlink2 = page.CreateHyperlink(text: "No, click this link", url: MyUrl);
hyperlink2.Click += SomeOtherMethod;
```

```cs
private int HandleTaskDialogCallback(IntPtr hWnd, Interop.TaskDialog.TASKDIALOG_NOTIFICATIONS notification, IntPtr wParam, IntPtr lParam)
{
    switch (notification)
    {
        case Interop.TaskDialog.TASKDIALOG_NOTIFICATIONS.TDN_HYPERLINK_CLICKED:
        {
            string hyperlink = Marshal.PtrToStringUni(lParam);  // hyperlink = "URL"
            TaskDialogHyperlinkClickedEventArgs eventArgs = new TaskDialogHyperlinkClickedEventArgs(hyperlink);
            
            TaskDialogHyperlink clickedHyperlink = ????   // what do we do here?
            clickedHyperlink.OnClick(eventArgs);

            break;
        }
    }
}
```

How would we handle this situation? Which handler are we suppose to invoke?

**Comment by weltkante** [comment]
> How would we handle this situation? Which handler are we suppose to invoke?

If the native control only delivers you the URL then you have to do the same thing you do in the RTF control, generate unique identifiers to lookup the original object. You would not pass on those identifiers, but take the URL of the object instead. The question is whether this is on the user or the framework to implement this workaround.

**Comment by jnm2** [comment]
I liked my `TextContent` API suggestion as the means of giving the page a reference to the hyperlinks, allowing `new TaskDialogHyperlink` instead of `page.CreateHyperlink`, because it felt more like how you usually create UI elements in Windows Forms: https://github.com/dotnet/winforms/issues/1199#issuecomment-691713029

Either way, one simple way the framework can determine which TaskDialogHyperlink instance gets its event raised is to emit its index within a collection of displayed links: `<a href="0">First link's text</a>, <a href="1">Second link's text</a>`. The framework would call `int.TryParse` when handling the notification to get the index.

**Comment by weltkante** [comment]
If you use indices as identifiers just make sure it plays well with changing the collection while the dialog is visible (or forbid it by throwing an exception)

**Comment by jnm2** [comment]
I already have a use case to change text while it's being shown that includes hyperlinks.

**Comment by RussKie** [comment]
> Either way, one simple way the framework can determine which TaskDialogHyperlink instance gets its event raised is to emit its index within a collection of displayed links as the href at the point of rendering to a text field for the native API:
> `<a href="0">First link's text</a>, <a href="1">Second link's text</a>`

I kind of feel we're overlooking the fact hyperlink type is a helper class, and it is not needed for hyperlink functionality - that is a user can set hyperlink text without using the helper. 
So it is not unexpected situation where a developer uses mix-and-match approach, where a text on a page is both set manually and with the hyperlink helper type. However, the results would be wildly different.

**Comment by RussKie** [comment]
We had another round of internal discussions, here's the outcome. Most of the code below is pseudocode.

## Objectives:
1. We need to support both scenarios:
    a. "unassisted": a developer builds a text unassisted, i.e.
    ```cs
    page.Text = $"Click my <a href="id">link</a>";
    ```
    b. "assisted": a developer uses our object model, i.e. 
    ```cs
    TaskDialogLink link1 = page.CreateLink(...);
    page.Text = $"Click my {link1}";
    ```
2. To accommodated the assisted scenario the object model must be both easily discoverable and familiar to a general Windows Forms developer.
3. Strive to avoid "hyperlink", "url", "uri", etc. terms to avoid creating incorrect expectations.

## Proposals

### A. Unassisted scenario

This is a valid scenario, though likely existing on the periphery of major use cases. It can be accommodated by exposing a `HyperlinkClicked` event at the page level, and allowing a developer to subclass the page to have ultimate control over the link functionality
```cs
// Windows Forms SDK code

public class TaskDialogPage
{
    protected internal virtual void OnHyperlinkClicked(TaskDialogHyperlinkClickedEventArgs e) { ... }
    
    protected internal bool EnableHyperlinks { get; set; }
}

public class TaskDialogDialog
{
    private int HandleTaskDialogCallback(IntPtr hWnd, Interop.TaskDialog.TASKDIALOG_NOTIFICATIONS notification, IntPtr wParam, IntPtr lParam)
    {
        switch (notification)
        {
            case Interop.TaskDialog.TASKDIALOG_NOTIFICATIONS.TDN_HYPERLINK_CLICKED:
            {
                string linkId = Marshal.PtrToStringUni(lParam);
                TaskDialogHyperlinkClickedEventArgs eventArgs = new TaskDialogHyperlinkClickedEventArgs(linkId);
                _boundPage.OnHyperlinkClicked(eventArgs);
                break;
            }
        }
    }
}
```

```cs
// Develop code

MyTaskDialogPage page = new (...);
page.Text = $"Click my <a href="id">link</a>";
page.EnableHyperlinks = true;


public class MyTaskDialogPage : TaskDialogPage
{
    protected override void OnHyperlinkClicked(TaskDialogHyperlinkClickedEventArgs e) 
    {
        // DO NOT CALL base.OnHyperlinkClicked(e);
        // Otherwise it will iterate over the links collection, and may invoke other handlers.
        
        // React to link clicks
    }
}
```

### Assisted scenario

This scenario is believed to be mostly used by Windows Forms developers. In this scenario hyperlinks are expected to be only added via the provided object model. 

:warning: Supplying a link via a raw string *may* render a link (if there is at least one added `TaskDialogLink`), however clicks on the raw link won't be handled.

```cs
// Windows Forms SDK code

public class TaskDialogLink
{
    public event EventHandler Click;
    
    internal TaskDialogLink(string text) 
    {
        Text = text;
        Id = Guid.NewGuid();
    }
    
    public string Text { get; }
    internal Guid Id { get; }
    internal void OnClick(EventArgs e) => Click?.Invoke(this, e);
    public override string ToString() => /* render <A HREF="Id">Text</A> with necessary escaping */
}
```

Since the provided link object has a `Click` event that must be handled, the only information is really necessary is the link text. We generate a unique identifier for the purpose of matching clicked links to their respective objects.

A `TaskDialogPage` maintains a list of hyperlinks in its private dictionary, so it can relay click events to required instance. To create an add a link to the collection it is proposed to use a factory method to reduce boilerplate code.


```cs
// Windows Forms SDK code

public class TaskDialogPage
{
    private Dictionary<Guid id, TaskDialogLink link> Links { get; } = new();
 
    public TaskDialogLink CreateLink(string text)
    {
        TaskDialogLink link = new (text);
        Links[link.Id] = link;
        EnableHyperLinks = true;
    }

    protected internal virtual void OnHyperlinkClicked(TaskDialogHyperlinkClickedEventArgs e)
    {
        TaskDialogLink link = Links[e.Id];
		link.OnClick(EventArgs.Empty);
    }
}

public class TaskDialogDialog
{
    private int HandleTaskDialogCallback(IntPtr hWnd, Interop.TaskDialog.TASKDIALOG_NOTIFICATIONS notification, IntPtr wParam, IntPtr lParam)
    {
        switch (notification)
        {
            case Interop.TaskDialog.TASKDIALOG_NOTIFICATIONS.TDN_HYPERLINK_CLICKED:
            {
                string linkId = Marshal.PtrToStringUni(lParam);
                TaskDialogHyperlinkClickedEventArgs eventArgs = new TaskDialogHyperlinkClickedEventArgs(linkId);
                _boundPage.OnHyperlinkClicked(eventArgs);
                break;
            }
        }
    }
}
```

```cs
// Develop code

TaskDialogLink hyperlink1 = page.CreateLink(text: "here");
hyperlink1.Click += (s, e) => Process.Start("https://github.com/dotnet/winforms/issues/1199");
TaskDialogLink hyperlink2 = page.CreateLink(text: "open new form");
hyperlink1.Click += (s, e) => new Form().Show();

TaskDialogPage page = new (...);
page.Text = $"Discuss task dialog hyperlink API proposal {hyperlink1}";
page.Footer.Text = $"You can {hyperlink2}!";
```

**Comment by KlausLoeffelmann** [comment]
Also, but that wasn't an discussion we had to the end, we decided we want to avoid the notion of being in the HTML realm altogether (since we're not), and so we want to avoid the term "Hyperlink" or "URL".

The latest suggestions, if I remember correctly, was `LinkLabel`, because that's pretty much what it is, and WinForms folks know it already.

**Comment by RussKie** [comment]
> Also, but that wasn't an discussion we had to the end, we decided we want to avoid the notion of being in the HTML realm altogether (since we're not), and so we want to avoid the term "Hyperlink" or "URL".

Yes, that's captured in Objective.3: _Strive to avoid "hyperlink", "url", "uri", etc. terms to avoid creating incorrect expectations_.




> The latest suggestions, if I remember correctly, was `LinkLabel`, because that's pretty much what it is, and WinForms folks know it already.

Do you mean like this?
```cs
public class TaskDialogLinkLabel
{
    public event EventHandler LinkClicked;
}
```
...to align with [`LinkLabel.LinkClicked`](https://docs.microsoft.com/dotnet/api/system.windows.forms.linklabel.linkclicked)?

**Comment by terrajobst** [comment]
* `LinkId` is a bit weird because ID implies some sort of opaque identifier that the developer have to map to something whereas it's really the value of the `href` attribute. If you want to avoid the web naming conventions, I suggest something like `LinkTarget`.
* It feels a bit heavy handed in the unassisted scenario having to derive from the page. It seems easier if `TaskDialogPage` would offer an event. Creating types is a fairly heavy operation and can't be done inside a method whereas an event can be subscribed to with a local function or lambda expression.
* We should either consistently use "hyperlink" or "link". We'd prefer "link" because "hyperlink" doesn't feel like UI but more like HTML.
* It seems you'll always need the unassisted model (for cases where the number of links isn't statically known). It also seems the result with factory methods wouldn't be easily toolable in the designer. It feels easier to just say the page/dialog has a `LinkClicked` event that provides the link target in the event args and not have `TaskDialogLink` class at all.

```C#
namespace System.Windows.Forms
{
    public class TaskDialogLink
    {
        public event EventHandler Click;
        public string Text { get; }
        public override string ToString();
    }
    public class TaskDialogLinkClickedEventArgs : EventArgs
    {
        public string LinkTarget { get; }
    }
    public partial class TaskDialogPage
    {
       protected bool EnableLinks { get; set; }
       public TaskDialogLink CreateLink(string text);
       protected virtual void OnLinkClicked(TaskDialogLinkClickedEventArgs e);
    }
}
```

**Comment by jnm2** [comment]
> It seems you'll always need the unassisted model (for cases where the number of links isn't statically known).

I didn't understand this until I watched the video, since you can create any number of links dynamically in the assisted model. What I picked up from the video is that the unassisted model is needed because a supported scenario is going to be loading preformed text including prerendered link tags from a source outside the app, such as from a database.

**Comment by jnm2** [comment]
I'm a little less comfortable with the term 'target' because that's an HTML concept too, and the difference in meaning is even greater. `Id` or `Identity` don't necessarily seem like things that have to be opaque to me.

I really like `Tag` for this though. It feels winformsy to me.

**Comment by jnm2** [comment]
Is there any interest in escaping the link text in case part of the text is constructed from user input? If the user types `</a> Hi <a href="something the app reacts to">`, I wonder if there is a security angle depending on how the app decides to act to the values. Hopefully apps don't handle `page.LinkClicked` using `Process.Start(e.LinkTarget);` as an easy shortcut for what they want to do.

**Comment by jnm2** [comment]
As an alternative to the `page.CreateLink` factory method, I still like the concept of `Text` always being escaped no matter what, and a `TextContent` collection property being available where you can choose whether you want to add text objects or link objects, suitable for tooling or for converting from your own object model, or whether you want to call `page.TextContent.Add($"Text {link} etc")` or similar as the normal case. Then links are always parented by the page's collection, so they can be created using `new TaskDialogLink`.

Both ways are also friendly for folks who like to use nested object initializers, whereas `page.CreateLink` is not friendly.

### Typical case

```cs
var aLink = new TaskDialogLink("a link");
aLink .Click += (_, _) => Foo();

var page = new TaskDialogPage
{
    Heading = "Example",
    TextContent = { $"Some text and {aLink}." },
}
```

### Tooling-friendly, more easy to convert from custom object model

```cs
var aLink = new TaskDialogLink("a link");
aLink .Click += (_, _) => Foo();

var page = new TaskDialogPage();
page.Heading = "Example";
page.TextContent.Add("Some text and ");
page.TextContent.Add(aLink);
page.TextContent.Add(".");
```

### Unassisted (prerendered links)

```cs
var page = new TaskDialogPage();
page.LinkClicked += (_, e) =>
{
    if (e.LinkTarget == "some expected tag") Foo();
};

page.Heading = "Example";
page.TextContent.Add("Some text and <a href=\"some expected tag\">a link</a>.");
```
```

**Comment by RussKie** [comment]
I went and tried implementing the proposal and looked at it from a developer side. And I do really feel we're trying to fit a square peg into a round hole. I'm going to rescind the current proposal for the following reasons:

1. Bloated API.
    All the previous examples assumed a page created, e.g.
    ```cs
    TaskDialogPage page = new (...);
    
    TaskDialogLink hyperlink1 = page.CreateLink(text: "here");
    hyperlink1.Click += (s, e) => Process.Start("https://github.com/dotnet/winforms/issues/1199");
    TaskDialogLink hyperlink2 = page.CreateLink(text: "open new form");
    hyperlink2.Click += (s, e) => new Form().Show();
    
    page.Text = $"Discuss task dialog hyperlink API proposal {hyperlink1}";
    page.Footer.Text = $"You can {hyperlink2}!";
    ```
    
    However, in reality a lot of developers will likely prefer the fluent expression, e.g.:
    https://github.com/dotnet/winforms/blob/e805f9253df702459c75a2961e7b7efd7026b2a4/src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/TaskDialogSamples.cs#L253-L274
    The current proposal makes it quite cumbersome - e.g. create a page instance, set all other properties except `Text`, then create links, and set `Text` properties.
    
    ```diff
    +            // THIS IS CUMBERSOME!
    +
                 // Show a task dialog (enhanced).
                 var page = new TaskDialogPage
                 {
                     Heading = "Are you sure you want to stop?",
    -                Text = "Stopping the operation might leave your database in a corrupted state.",
    +                // We can't create and reference links here :(
    +                // Text = ????
                     Caption = "Confirmation (Task Dialog)",
                     Icon = TaskDialogIcon.Warning,
                     AllowCancel = true,
     
                     Verification = new TaskDialogVerificationCheckBox
                     {
    -                    Text = "Do not show again"
    +                    Text = "Do not show again" // doesn't support hyperlinks
                     },
    +
    +                Expander = new TaskDialogExpander
    +                {
    +                    Text = "Do not show again"
    +                    Position = TaskDialogExpanderPosition.AfterText
    +
    +                    // We can't create and reference links here :(
    +                    // Text = ????
    +                },
    +
    +                Footnote = new TaskDialogFootnote
    +                {
    +                    Icon = TaskDialogIcon.Information
    +
    +                    // We can't create and reference links here :(
    +                    // Text = ????
    +                },
    +
    
    +            TaskDialogLink link1 = page.CreateLink("Click me!");
    +            link1.Click += (s, e) => Debug.WriteLine($"{nameof(link1)} was clicked");
    +            page.Text = $"Stopping the operation might leave your database in a corrupted state. {link1}";
    +
    +            TaskDialogLink link2 = page.CreateLink("Click me!");
    +            link2.Click += (s, e) => Debug.WriteLine($"{nameof(link2)} was clicked");
    +            page.Expander.Text = $"No! {link2}";
    +
    +            TaskDialogLink link3 = page.CreateLink("Click me!");
    +            link3.Click += (s, e) => Debug.WriteLine($"{nameof(link3)} was clicked");
    +            page.Footnote.Text = $"{link3} in the footer";
    +
    +            // END THIS
    ```
    This goes against our aim to make API as user-friendly and as little boilerplate as possible.

2. As raised during the API Review, the current proposal feels "heavy handed" for use-cases where links are provided by developers (i.e. encoded as HTML and not created via `page.CreateLink()` method). 
Besides, manually encoded links will need to be handled at the `page` level (i.e. `page.OnLinkClicked(TaskDialogLinkClickedEventArgs)`) and not at the "link" level. This maintains the duality that we wanted to avoid in the first place.

3. The current proposal doesn't work for "hybrid" scenarios where some links are created via `page.CreateLink()` method and some links are manually encoded.

4. The proposal does not really have a clear benefit for the design story. In fact, IMO, it makes it harder to integrate links into `Text` properties.

Whilst we can/try to think of links as distinct "components" (in Windows Forms sense), it is equally reasonable to think of links as being a part of textual information on a page with an added click-ability.
If we follow the latter model - i.e. a "click on a link" is similar to "click on a page", handling a click event at the page level is not totally horrible. And we offer a reasonably straight forward the following developer experience:
```cs
            string linkId1 = Guid.NewGuid().ToString();
            string linkId2 = Guid.NewGuid().ToString();
            string linkId3 = "https://localhost/";

            // Show a task dialog (enhanced).
            var page = new TaskDialogPage
            {
                EnableHyperLinks = true,
                Text = $"Stopping the operation might leave your database in a corrupted state. {TaskDialogPage.CreateLink(linkId1, "Click me!")}",

                Expander = new TaskDialogExpander
                {
                    Text = TaskDialogPage.CreateLink(linkId2, "Click me!")
                },

                Footnote = new TaskDialogFootnote
                {
                    Text = $"{TaskDialogPage.CreateLink(linkId3, "Click me!")} in the footer",
                },

            };

            page.LinkClicked += (s, e) =>
            {
                if (e.LinkTarget == linkId1)
                {
                    Debug.WriteLine($"Clicked {nameof(linkId1)} -> {e.LinkTarget}");
                }
                else if (e.LinkTarget == linkId2)
                {
                    Debug.WriteLine($"Clicked {nameof(linkId2)} -> {e.LinkTarget}");
                }
            };
```

This is very much what the original proposal from @kpreisser was: https://github.com/kpreisser/winforms/commit/187bdb4ecb8d13296a3cedc617470f2115fffc95.

Yes, the API discoverability is probably a little vague, but I'm sure, having read the docs on how to use the feature once a developer shouldn't have any troubles to use the feature.

. . .

`TaskDialogPage.CreateLink` represents something that @GSPP and @jnm2 have (in some shape or form) proposed earlier:
```cs
        public static string CreateLink(string target, string text)
        {
            return $"<A HREF=\"{target}\">{text}</A>"; /* render <A HREF="Id">Text</A> with necessary escaping */
        }
```
I'm not totally happy with the name and the location of this helper method; and it is certainly open for further discussion.

**Comment by KlausLoeffelmann** [comment]
> <a href=\"{linkId1}\">Click me!</a>}"

I find this absolutely dreaded, horrifying and terrible - don't get mad yet, this sentence has not finished 😄 - in the _context of the original WinForms Development spirit!_

> but I'm sure, having read the docs on how to use the feature once a developer shouldn't have any troubles to use the feature.

Read the docs? Come on...
It is completely undiscoverable, doesn't aim our target audience, and I predict, it will not be widely accepted.

> As raised during the API Review, the current proposal feels "heavy handed" for use-cases where links are provided by developers...

Please, let's not assume that clicking a Link like www.learnhtmlsoyoucanusewinforms.com would automatically start the browser just by clicking on it. At least that's not how I understand it. So, I guess what I am saying is: the additional work with reflecting everything with an object model is not much more work in contrast to what you actually need to do in addition, to make the mentioned scenario work (start the browser process, remote-control resolving the link, showing it in the browser, etc.). On top, it's also easier to understand what's happening, when you read the code.

So, to wrap my remarks up: This in the current suggestion-state would be an API for a few. Maybe something which you'd like to use, @terrajobst like to use, @weltkante would like to use or even I would use, because I was part of the Design discussion, so I should know how to use it (but will I in 2 years from now?) -  *but my really big concern is: _not_ for the typical WinForms Dev.*

As a proxy for a dev of our typical target audience, I am really, really against this, and like us all to come up with a compromise which has a way better discoverability and no need to read the docs, but that every Level 100 WinForms Dev should understand intuitively.

My 2 cents.

**Comment by jnm2** [comment]
> `/// Gets the value of the <c>target</c> attribute of the link that the user clicked.`

Should say the href attribute, right?

Since target is an HTML concept that means something, as in `<a href="..." target="_blank">`, can we go with `LinkId` instead of `LinkTarget`?

**Comment by RussKie** [comment]
We've had a chat with @KlausLoeffelmann and he will put forward another proposal.

**Comment by tannergooding** [comment]
My 2 cents here is that what @RussKie proposed (https://github.com/dotnet/winforms/issues/1199#issuecomment-787822157) is a small minimal abstraction that fits well to the Win32 APIs actually exposed.
`public bool EnableLinks { get; set; }` maps to [TDF_ENABLE_HYPERLINKS](https://docs.microsoft.com/en-us/windows/win32/api/commctrl/ns-commctrl-taskdialogconfig#members) while `public event EventHandler<TaskDialogLinkClickedEventArgs>? LinkClicked;` corresponds to the [TDN_HYPERLINK_CLICKED](https://docs.microsoft.com/en-us/windows/win32/controls/tdn-hyperlink-clicked) notification code.

The other abstractions, like what @KlausLoeffelmann has expressed interest in should be buildable on top of this and exposed for users who do need or desire it while the reverse is not true.

The only thing actually missing here is a helper API to take text and insert the hyperlinks. An API such as was listed in the most recent proposal handles this:
```csharp
public static string CreateLink(string target, string text)
{
    return $"<A HREF=\"{target}\">{text}</A>"; /* render <A HREF="Id">Text</A> with necessary escaping */
}
```

Such an API handles creating the relevant anchor tag which you could then use with `string.Format` to actually fill in your relevant text. It's allows what is basically a one liner and is fairly cheap. That is, it incurs a minimal number of allocations and represents how these actually work and so it allows other abstractions to be easily built on top as appropriate.

Given the above, you can even add a helper API such as `public static string InsertLinks(string format, params (string target, string text) links)` (and/or overloads for 1/2/3 links) which can internally use `string builder` and simplifies the usage even further, all without adding additional abstractions over the links themselves.

In both scenarios, such an API can handle any necessary escaping and simplifies the overall design when this is being data driven while still allowing users to build their own abstractions if they feel that suits their needs.

**Comment by tannergooding** [comment]
You can always build a higher abstraction over a low-cost minimal wrapper. The reverse is rarely true.

**Comment by KlausLoeffelmann** [comment]
If I remember correctly, the original proposal provided this.
I have no idea, why we did not pursue the original draft? 
What was the problem with the original proposal, where we already had the Unassisted Scenario (see above)?

Please let's not forget: The TaskDialog or a Page of it is a modal Dialogbox which most likely is dedicatedly crafted for a typical UI scenario anyway. It is not the result of an aggregation of data, or an result set of links, which results in hundreds of text bodies to be displayed. 

The most likely scenario is, that a Developer is displaying a fixed, completely undynamically created Dialogbox based on a specific UI context. And for that, discoverability of functionality is way more important than everything else. Also, the memory consumption of the proposed object model is absolutely neglectable, here. We're not creating pages in a loop. We are not showing modal DialogBoxes (so TaskDialogs) by the hundreds! The typical scenario creates maybe 5 pages in the TaskDialog, shows the dialog, and be done. Other scenarios are niche scenarios, and for that we already had the Unassisted Scenario.

**Comment by tannergooding** [comment]
My general point here is that the proposal given [here](https://github.com/dotnet/winforms/issues/1199#issuecomment-787822157) represents the minimum viable product. It exposes all the underlying functionality supported by the API and allows all the scenarios, both expected and unexpected. It allows you to be as performant or allocating as you want and you can build all other functionalities on top of it. Any other features, such as an assisted scenario are effectively built on top of this MVP.

There should not be an API proposal, IMO, that does not involve exposing this core functionality.

If there is some assisted scenario that does work well on top of the minimum API set, I think that is reasonable, but I strongly believe it should be built on top of it.
That could be in box, but the MVP API discussed means that it could also be provided via a 3rd party with no additional issues.

**Comment by KlausLoeffelmann** [comment]
I understood that the unassisted scenario is the base, and the more guided or rather guiding API is on top.
Will talk to @RussKie  again about this, and review! Thanks!

**Comment by KlausLoeffelmann** [comment]
Guys, I'll need some time to think about this, and try to come up with some additional ideas around this. 
Sorry for that, but I am not forgetting it.

**Comment by merriemcgaw** [comment]
I tend to agree with @tannergooding 's comment [here](https://github.com/dotnet/winforms/issues/1199#issuecomment-788401639) and that given where we are in the cycle, and what resources we have available, I think we should start with what we have here, and a minimum viable product for the unassisted scenarios. If we get feedback about this specific part of the API, we can always add layers of abstraction over it. I'm OK with the proposal the way it stands today. This doesn't have to reach all of the Level 100 WinForms developers, in fact I'd argue that the unassisted scenario is specifically NOT targeted at these developers. That's where the easy option comes in.

**Comment by RussKie** [comment]
We're proceeding with the following API, which is very much what the original proposal from @kpreisser was: kpreisser@187bdb4.

I acknowledge the API discoverability is a little vague, but having read the docs once on how to use the feature developers shouldn't have any troubles using it again. 
We currently won't be providing any helper methods to construct HREFs.

We can revisit the API and any helper functionality, if/when we detect a significant uptake of this API.

### Proposed API

<details>
<summary>Details</summary>

```diff
+    /// <summary>
+    /// Provides data for the <see cref="TaskDialogPage.LinkClicked"/> event.
+    /// </summary>
+    public class TaskDialogLinkClickedEventArgs : EventArgs
+    {
+        /// <summary>
+        /// Gets the value of the <c>target</c> attribute of the link that the user clicked.
+        /// </summary>
+        /// <remarks>
+        /// Note: In order to avoid possible security vulnerabilities when showing content
+        /// from unsafe sources in a task dialog, you should always verify the value of this
+        /// property before actually opening the link.
+        /// </remarks>
+        public string LinkTarget { get; }
+    }

     public class TaskDialogPage
     {
+        /// <summary>
+        ///   Occurs when the user has clicked on a link.
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        ///   This event will only be raised if <see cref="EnableLinks"/> is set to <see langword="true"/>.
+        /// </para>
+        /// </remarks>
+        public event EventHandler<TaskDialogLinkClickedEventArgs>? LinkClicked;
+
+        /// <summary>
+        /// <para>
+        ///   Gets or sets a value that specifies whether the task dialog should
+        ///   interpret strings in the form <c>&lt;a href="target"&gt;link Text&lt;/a&gt;</c>
+        ///   as hyperlink when specified in the <see cref="Text"/>,
+        ///   <see cref="TaskDialogFooter.Text"/>
+        ///   or <see cref="TaskDialogExpander.Text"/> 
+        ///   or <see cref="TaskDialogFootnote.Text"/> properties.
+        ///   When the user clicks on such a link, the <see cref="LinkClicked"/>
+        ///   event is raised, containing the value of the <c>target</c> attribute.
+        /// </para>
+        /// </summary>
+        /// <value>
+        ///   <see langword="true"/> to enable links; otherwise, <see langword="false"/>.
+        ///   The default value is <see langword="false"/>.
+        /// </value>
+        /// <remarks>
+        /// <para>
+        ///   The Task Dialog will not actually execute any links.
+        ///   Link execution must be handled in the <see cref="LinkClicked"/> event.
+        /// </para>
+        /// <para>
+        ///   Note: Enabling this setting causes the <c>"&amp;"</c> character to be
+        ///   interpreted as prefix for an access key character (mnemonic) if at least
+        ///   one link is used.
+        /// </para>
+        /// <para>
+        ///   When you enable this setting and you want to display a text
+        ///   without interpreting links, you must replace the strings <c>"&lt;a"</c>
+        ///   and <c>"&lt;A"</c> with something like <c>"&lt;\u200Ba"</c>.
+        /// </para>
+        /// </remarks>
+        public bool EnableLinks { get; set; }
     }
```


</details>

**Comment by KlausLoeffelmann** [comment]
OK then, I agree meekly. 😄 (And sorry for not getting to this sooner, I am still swamped with higher priority Designer and VB things). I still think we should have something in edition, but this API doesn't prevent us from extensions in the future, so I completely agree!

**Comment by jnm2** [comment]
Are PRs currently welcome for this from outside contributors?

**Comment by dreddy-work** [comment]
> Are PRs currently welcome for this from outside contributors?

Yes. Please feel free to send PR. Label `help wanted` is specifically designed for this.

**Comment by RussKie** [comment]
I was planning to complete this story but there was something else with higher priority. I have an old branch https://github.com/dotnet/winforms/compare/main...RussKie:winforms:fix_1199, which I _think_ has the latest implementation. Feel free to reuse it, if you find it useful.
Also, please have a look at the original @kpreisser's implementation in https://github.com/kpreisser/winforms/commit/187bdb4.

**Comment by jnm2** [comment]
Should the event args property not be called LinkHref? There's a consistent mistake in the terminology of the XML docs where 'target' is used instead of 'href':

```diff
 /// <summary>
 /// Provides data for the <see cref="TaskDialogPage.LinkClicked"/> event.
 /// </summary>
 public class TaskDialogLinkClickedEventArgs : EventArgs
 {
     /// <summary>
-    /// Gets the value of the <c>target</c> attribute of the link that the user clicked.
+    /// Gets the value of the <c>href</c> attribute of the link that the user clicked.
     /// </summary>
     /// <remarks>
     /// Note: In order to avoid possible security vulnerabilities when showing content
     /// from unsafe sources in a task dialog, you should always verify the value of this
     /// property before actually opening the link.
     /// </remarks>
-    public string LinkTarget { get; }
-    public string LinkHref { get; }
 }
```

```diff
 ///   When the user clicks on such a link, the <see cref="LinkClicked"/>
-///   event is raised, containing the value of the <c>target</c> attribute.
+///   event is raised, containing the value of the <c>target</c> attribute.
```

**Comment by RussKie** [comment]
Sounds good.

**Comment by jnm2** [comment]
Ready for review at #9056!

**Comment by Amy-Li03** [comment]
Verified this on .NET 8.0.100-preview.5.23253.14 using Task Dialog demo, it works well, following video is the test result.
![Fixed](https://user-images.githubusercontent.com/62929087/236122706-2ff5dae0-643d-422f-9ced-3b31c942851d.gif)

**Comment by Olina-Zhang** [comment]
Verified this on .NET 8.0 test pass build: 8.0.100-preview.5.23276.4, issue was fixed, test result is same as above video.
